waku/waku_filter_v2/subscriptions.nim

```diff
@@ -6,48 +6,161 @@ else:
 import
   std/[sets,tables],
   chronicles,
-  libp2p/peerid
+  chronos,
+  libp2p/peerid,
+  stew/shims/sets
 import
-  ../waku_core
+  ../waku_core,
+  ../utils/tableutils
 
 logScope:
   topics = "waku filter subscriptions"
 
 const
-  MaxTotalSubscriptions* = 1000 # TODO make configurable
-  MaxCriteriaPerSubscription* = 1000
+  MaxFilterPeers* = 1000
+  MaxFilterCriteriaPerPeer* = 1000
+  DefaultSubscriptionTimeToLiveSec* = 5.minutes
 
 type
-  FilterCriterion* = (PubsubTopic, ContentTopic) # a single filter criterion is fully defined by a pubsub topic and content topic
+  # a single filter criterion is fully defined by a pubsub topic and content topic
+  FilterCriterion* = tuple
+    pubsubTopic: PubsubTopic
+    contentTopic: ContentTopic
+
   FilterCriteria* = HashSet[FilterCriterion] # a sequence of filter criteria
-  FilterSubscriptions* = Table[PeerID, FilterCriteria] # a mapping of peer ids to a sequence of filter criteria
 
-proc findSubscribedPeers*(subscriptions: FilterSubscriptions, pubsubTopic: PubsubTopic, contentTopic: ContentTopic): seq[PeerID] =
-  ## Find all peers subscribed to a given topic and content topic
-  let filterCriterion = (pubsubTopic, contentTopic)
+  SubscribedPeers* = HashSet[PeerID] # a sequence of peer ids
+
+  PeerData* = tuple
+    lastSeen: Moment
+    criteriaCount: uint
+
+  FilterSubscriptions* = object
+    peersSubscribed*    : Table[PeerID, PeerData]
+    subscriptions*      : Table[FilterCriterion, SubscribedPeers]
+    subscriptionTimeout*: Duration
+    maxPeers*           : uint
+    maxCriteriaPerPeer* : uint
+
+proc init*(T: type FilterSubscriptions,
+          subscriptionTimeout: Duration = DefaultSubscriptionTimeToLiveSec,
+          maxFilterPeers: uint32 = MaxFilterPeers,
+          maxFilterCriteriaPerPeer: uint32 = MaxFilterCriteriaPerPeer): FilterSubscriptions =
+  ## Create a new filter subscription object
+  return FilterSubscriptions(
+              peersSubscribed: initTable[PeerID, PeerData](),
+              subscriptions: initTable[FilterCriterion, SubscribedPeers](),
+              subscriptionTimeout: subscriptionTimeout,
+              maxPeers: maxFilterPeers,
+              maxCriteriaPerPeer: maxFilterCriteriaPerPeer
+            )
+
+proc isSubscribed*(s: var FilterSubscriptions, peerId: PeerID): bool =
+  s.peersSubscribed.withValue(peerId, data):
+    return Moment.now() - data.lastSeen <= s.subscriptionTimeout
+
+  return false
+
+proc subscribedPeerCount*(s: FilterSubscriptions): uint =
+  return cast[uint](s.peersSubscribed.len())
```

Comment:
Tiny detail. When using a "noun", better use it without parenthesis. Applies elsewhere.
```suggestion
  return cast[uint](s.peersSubscribed.len)
```